highlight/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+            ranges,
+            at_node_end: false,
+        }
+    }
+
+    fn offset(&self) -> usize {
+        if self.at_node_end {
+            self.cursor.node().end_byte()
+        } else {
+            self.cursor.node().start_byte()
+        }
+    }
+
+    fn advance(&mut self) -> bool {
+        if self.at_node_end {
+            if self.cursor.goto_next_sibling() {
+                self.at_node_end = false;
+            } else if !self.cursor.goto_parent() {
+                return false;
+            }
+        } else if !self.cursor.goto_first_child() {
+            self.at_node_end = true;
+        }
+        true
+    }
+}
+
+impl<'de> Deserialize<'de> for Scope {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        match s.as_str() {
+            "attribute" => Ok(Scope::Attribute),
+            "comment" => Ok(Scope::Comment),
+            "constant" => Ok(Scope::Constant),
+            "constant.builtin" => Ok(Scope::ConstantBuiltin),
+            "constructor" => Ok(Scope::Constructor),
+            "constructor.builtin" => Ok(Scope::ConstructorBuiltin),
+            "embedded" => Ok(Scope::Embedded),
+            "escape" => Ok(Scope::Escape),
+            "function" => Ok(Scope::Function),
+            "function.builtin" => Ok(Scope::FunctionBuiltin),
+            "keyword" => Ok(Scope::Keyword),
+            "number" => Ok(Scope::Number),
+            "operator" => Ok(Scope::Operator),
+            "property" => Ok(Scope::Property),
+            "property.builtin" => Ok(Scope::PropertyBuiltin),
+            "punctuation" => Ok(Scope::Punctuation),
+            "punctuation.bracket" => Ok(Scope::PunctuationBracket),
+            "punctuation.delimiter" => Ok(Scope::PunctuationDelimiter),
+            "punctuation.special" => Ok(Scope::PunctuationSpecial),
+            "string" => Ok(Scope::String),
+            "string.special" => Ok(Scope::StringSpecial),
+            "type" => Ok(Scope::Type),
+            "type.builtin" => Ok(Scope::TypeBuiltin),
+            "variable" => Ok(Scope::Variable),
+            "variable.builtin" => Ok(Scope::VariableBuiltin),
+            "tag" => Ok(Scope::Tag),
+            _ => Ok(Scope::Unknown),
+        }
+    }
+}
+
+pub trait HTMLAttributeCallback<'a>: Fn(Scope) -> &'a str {}
+
+pub fn highlight<'a, F>(
+    source: &'a [u8],
+    language: Language,
+    property_sheet: &'a PropertySheet<Properties>,
+    injection_callback: &'a F,
+) -> Result<impl Iterator<Item = HighlightEvent<'a>> + 'a, String>
+where
+    F: Fn(&str) -> Option<(Language, &'a PropertySheet<Properties>)>,
+{
+    Highlighter::new(source, language, property_sheet, injection_callback)
+}
+
+pub fn highlight_html<'a, F1, F2>(
+    source: &'a [u8],
+    language: Language,
+    property_sheet: &'a PropertySheet<Properties>,
+    injection_callback: &'a F1,
+    attribute_callback: &'a F2,
```

Comment:
The `highlight_html` function also takes a second callback - `attribute_callback` that specifies how to render a scope as HTML. In the case of the CLI, it uses an inline `style` attribute, whereas other apps may want to map scopes to their own classes.